src/main/kotlin/org/lsst/dax/albuquery/resources/Async.kt

```diff
@@ -77,37 +79,102 @@ class Async(val metaservDAO: MetaservDAO) {
         val query = queryParam ?: postBody
+        val format = formatParam ?: ""
         LOGGER.info("Recieved query [$query]")
-        val mapper: ObjectMapper
+        var mapper: ObjectMapper? = null
         val ct = headers.getRequestHeader(HttpHeaders.ACCEPT).get(0)
-        if (ct == MediaType.APPLICATION_XML)
-            mapper = TableMapper()
-        else // JSON as default
+        if (ct == MediaType.APPLICATION_JSON || format.contains("json"))
             mapper = ObjectMapper().registerModule(KotlinModule())
+        if (mapper == null) // default
+            mapper = TableMapper()
         return createAsyncQuery(metaservDAO, uri, query, mapper, true)
     }
 
     @Timed
     @GET
-    @Path("{id}/results/{result}")
-    @Produces(MediaType.APPLICATION_JSON, MediaType.APPLICATION_XML)
-    fun getQuery(@PathParam("id") queryId: String, @PathParam("result") result: String): Response {
+    @Path("{id}{p: (/parameters)?}")
+    @Produces(MediaType.APPLICATION_JSON)
+    fun getQueryParams(@PathParam("id") queryId: String): Response {
+        val queryTaskFuture = findOutstandingQuery(queryId)
+        if (queryTaskFuture != null) {
+            val queryTask = queryTaskFuture.get(1, TimeUnit.MILLISECONDS)
+            val params = queryTask.phaseInfo.parameters
+            return Response.ok("Query=" + params).build()
+        }
+        else
+            return Response.status(Response.Status.NOT_FOUND).build()
+    }
+
+    @Timed
+    @GET
+    @Path("{id}/error")
+    @Produces(MediaType.APPLICATION_JSON)
+    fun getQueryStatus(@PathParam("id") queryId: String): Response {
+        val queryTaskFuture = findOutstandingQuery(queryId)
+        if (queryTaskFuture != null) {
+            val queryTask = queryTaskFuture.get(1, TimeUnit.MILLISECONDS)
+            val errorFile = queryTask.phaseInfo.error
+            return Response.ok(errorFile).build()
+        }
+        else
+            return Response.status(Response.Status.NOT_FOUND).build()
+
+    }
+
+    @Timed
+    @GET
+    @Path("{id}/results")
+    @Produces(MediaType.APPLICATION_JSON)
+    fun getQueryResults(@PathParam("id") queryId: String): Response {
+        val queryTaskFuture = findOutstandingQuery(queryId)
+        if (queryTaskFuture != null) {
+            if (queryTaskFuture.isDone()) {
+                val result = getResultFile(queryId, "result")
+                return Response.ok(result).build()
+            }
+            else {
+                    val resultUri = getResultUri(uri, queryId, true)
+                    return Response.seeOther(resultUri).build()
+                }
+        }
+        else
+            return Response.status(Response.Status.NOT_FOUND).build()
+    }
+
+    @Timed
+    @GET
+    @Path("{id}/results/result")
+    @Produces( MediaType.APPLICATION_XML, MediaType.APPLICATION_JSON )
+    fun getQuery(
+        @PathParam("id") queryId: String,
+        @QueryParam("RESPONSEFORMAT") formatParam: String?
```

Comment:
So just to be clear here, you have to send RESPONSEFORMAT at the creation, and then send it as well for getting the result?  I was under the impression that whatever the format was set to as a parameter when the job was run determined this.  Either way is fine with me as long as it doesn't break anyone.

Reply:
RESPONSEFORMAT is the VO way to specify the preferred document format in response. The PhaseInfo.format parameter is used to keep track of this.

Reply:
Right, I get that, and that makes sense, that's the spec.

My point is that RESPONSEFORMAT Is only sent when creating the job.  Then when you get the results, it should reuse the RESPONSEFORMAT that was made when creating the job.

This might be my lack of kotlin showing through, but where do you assign the format you get to the format parameter of phaseinfo?  I don't see any line where phaseinfo.format is assigned.

Reply:
The phaseinfo.format is assined in Async.createAsyncQuery() in housekeeping section.

Reply:
Ah I see, although this is a bit confusing, because you have the format there, but then you use it to create an object mapper, then infer the type back again when you look at what type the mapper is.  This is fine I suppose, although maybe I'd just pass in what format you think it is, then create the object mapper inside?

Reply:
This is an area that we'll have to revisit, as SUIT has pointed out, we want to support non-VO format(s) as well.  For now, we have the XML (VO) and JSON (proprietary LSST) mappers.  If the incoming format parameters do not go into either one, then they're invalid.  Perhaps there is clearer to say this in the future.